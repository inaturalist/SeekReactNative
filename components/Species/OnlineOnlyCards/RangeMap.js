// @flow
import React, { useState, useEffect } from "react";
import { Image, TouchableOpacity } from "react-native";
import MapView, { PROVIDER_DEFAULT, UrlTile, Marker } from "react-native-maps";
import { useRoute, useNavigation } from "@react-navigation/native";
import type { Node } from "react";

import i18n from "../../../i18n";
import { viewStyles, textStyles } from "../../../styles/species/rangeMap";
import { fetchTruncatedUserLocation } from "../../../utility/locationHelpers";
import icons from "../../../assets/icons";
import Legend from "../../Modals/LegendModal";
import Modal from "../../UIComponents/Modals/Modal";
import ViewWithHeader from "../../UIComponents/Screens/ViewWithHeader";
import StyledText from "../../UIComponents/StyledText";

const latitudeDelta = 0.2;
const longitudeDelta = 0.2;

const RangeMap = (): Node => {
  const navigation = useNavigation();
  const { params } = useRoute();
  // region can be the obs region or the user location, depending on nav
  const { region, id, seenDate } = params;

  const [showModal, setModal] = useState( false );
  const [user, setUser] = useState( {} );
  const [mapRegion, setMapRegion] = useState( region );

  const openModal = () => setModal( true );
  const closeModal = () => setModal( false );

  const getUserLocation = () => {
    fetchTruncatedUserLocation().then( ( coords ) => {
      if ( coords ) {
        const { latitude, longitude } = coords;

        setUser( {
          latitude,
          longitude,
          latitudeDelta,
          longitudeDelta
        } );
      }
    } ).catch( ( error ) => {
      if ( error ) {
        setUser( {} );
      }
    } );
  };

  const updateMap = () => {
    // only show userLocation button if permissions are on
    // a user can have location off and still see range map for previous observation locations
    if ( !user.latitude ) {
      return;
    }

    setMapRegion( {
      latitude: user.latitude,
      longitude: user.longitude,
      latitudeDelta,
      longitudeDelta
    } );
  };

  useEffect( () => {
    navigation.addListener( "focus", () => {
      getUserLocation();
    } );
  }, [navigation] );

  if ( mapRegion.latitude === undefined ) {
    return null;
  }

  return (
    <ViewWithHeader testID="range-map-container" header="species_detail.range_map" footer={false}>
      <Modal
        showModal={showModal}
        closeModal={closeModal}
        modal={<Legend closeModal={closeModal} />}
      />
      <MapView
        testID="range-map"
        provider={PROVIDER_DEFAULT}
        region={mapRegion}
        onRegionChangeComplete={setMapRegion}
        style={viewStyles.map}
        zoomEnabled
      >
        <UrlTile
          tileSize={512}
          urlTemplate={`https://api.inaturalist.org/v1/grid/{z}/{x}/{y}.png?taxon_id=${id}&color=%2377B300&verifiable=true`}
        />
        {seenDate && (
          <Marker coordinate={{ latitude: region.latitude, longitude: region.longitude }}>
            <Image source={icons.cameraOnMap} />
          </Marker>
        )}
        {user.latitude && (
          <Marker coordinate={{ latitude: user.latitude, longitude: user.longitude }}>
            <Image source={icons.locationPin} />
          </Marker>
        )}
      </MapView>
      <TouchableOpacity
        onPress={openModal}
        style={viewStyles.legend}
      >
        <StyledText style={textStyles.whiteText}>
          {i18n.t( "species_detail.legend" ).toLocaleUpperCase()}
        </StyledText>
      </TouchableOpacity>
      {user.latitude && (
        <TouchableOpacity
          accessibilityLabel={i18n.t( "accessibility.user_location" )}
          accessible
          testID="user-location-button"
          onPress={updateMap}
          style={viewStyles.locationIcon}
        >
          <Image source={icons.indicator} />
        </TouchableOpacity>
      )}
    </ViewWithHeader>
  );
};

export default RangeMap;
