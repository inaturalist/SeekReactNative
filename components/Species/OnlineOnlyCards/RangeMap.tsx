import React, { useState, useEffect } from "react";
import { Image, TouchableOpacity } from "react-native";
import MapView, { PROVIDER_DEFAULT, UrlTile, Marker } from "react-native-maps";
import { useRoute, useNavigation } from "@react-navigation/native";

import i18n from "../../../i18n";
import { viewStyles, textStyles } from "../../../styles/species/rangeMap";
import { fetchTruncatedUserLocation } from "../../../utility/locationHelpers";
import icons from "../../../assets/icons";
import Legend from "../../Modals/LegendModal";
import Modal from "../../UIComponents/Modals/Modal";
import ViewWithHeader from "../../UIComponents/Screens/ViewWithHeader";
import StyledText from "../../UIComponents/StyledText";
import { baseTextStyles } from "../../../styles/textStyles";

const latitudeDelta = 0.2;
const longitudeDelta = 0.2;

const RangeMap = () => {
  const navigation = useNavigation();
  const { params } = useRoute();
  // TODO: navigation TS
  // region can be the obs region or the user location, depending on nav
  const { region, id, seenDate } = params;

  const [showModal, setModal] = useState( false );
  const [user, setUser] = useState<{
    latitude?: number;
    longitude?: number;
    latitudeDelta?: number;
    longitudeDelta?: number;
  }>( {} );
  const [mapRegion, setMapRegion] = useState<{
    latitude: number;
    longitude: number;
    latitudeDelta: number;
    longitudeDelta: number;
  }>( region );

  const openModal = () => setModal( true );
  const closeModal = () => setModal( false );

  const getUserLocation = () => {
    fetchTruncatedUserLocation().then( ( coords ) => {
      if ( coords ) {
        const { latitude, longitude } = coords;

        setUser( {
          latitude,
          longitude,
          latitudeDelta,
          longitudeDelta
        } );
      }
    } ).catch( ( error ) => {
      if ( error ) {
        setUser( {} );
      }
    } );
  };

  const updateMap = () => {
    // only show userLocation button if permissions are on
    // a user can have location off and still see range map for previous observation locations
    if ( !user.latitude || !user.longitude ) {
      return;
    }

    setMapRegion( {
      latitude: user.latitude,
      longitude: user.longitude,
      latitudeDelta,
      longitudeDelta
    } );
  };

  useEffect( () => {
    const unsubscribe = navigation.addListener( "focus", () => {
      getUserLocation();
    } );

    return unsubscribe;
  }, [navigation] );

  if ( mapRegion.latitude === undefined ) {
    return null;
  }

  return (
    <ViewWithHeader testID="range-map-container" header="species_detail.range_map" footer={false}>
      <Modal
        showModal={showModal}
        closeModal={closeModal}
        modal={<Legend closeModal={closeModal} />}
      />
      <MapView
        testID="range-map"
        provider={PROVIDER_DEFAULT}
        region={mapRegion}
        onRegionChangeComplete={setMapRegion}
        style={viewStyles.map}
        zoomEnabled
      >
        <UrlTile
          tileSize={512}
          urlTemplate={`https://api.inaturalist.org/v1/grid/{z}/{x}/{y}.png?taxon_id=${id}&color=%2377B300&verifiable=true`}
        />
        {seenDate && (
          <Marker coordinate={{ latitude: region.latitude, longitude: region.longitude }}>
            <Image source={icons.cameraOnMap} />
          </Marker>
        )}
        {user.latitude && user.longitude && (
          <Marker coordinate={{ latitude: user.latitude, longitude: user.longitude }}>
            <Image source={icons.locationPin} />
          </Marker>
        )}
      </MapView>
      <TouchableOpacity
        onPress={openModal}
        style={viewStyles.legend}
      >
        <StyledText style={[baseTextStyles.modalBanner, textStyles.whiteText]}>
          {i18n.t( "species_detail.legend" ).toLocaleUpperCase()}
        </StyledText>
      </TouchableOpacity>
      {user.latitude && (
        <TouchableOpacity
          accessibilityLabel={i18n.t( "accessibility.user_location" )}
          accessible
          testID="user-location-button"
          onPress={updateMap}
          style={viewStyles.locationIcon}
        >
          <Image source={icons.indicator} />
        </TouchableOpacity>
      )}
    </ViewWithHeader>
  );
};

export default RangeMap;
