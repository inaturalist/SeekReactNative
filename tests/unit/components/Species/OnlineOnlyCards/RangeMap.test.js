import React from "react";
import { render, screen } from "@testing-library/react-native";

import RangeMap from "../../../../../components/Species/OnlineOnlyCards/RangeMap";

jest.mock( "@react-navigation/native", () => {
  const actualNav = jest.requireActual( "@react-navigation/native" );
  return {
    ...actualNav,
    useNavigation: () => ( {
      navigate: jest.fn(),
      dispatch: jest.fn(),
      // To intercept on focus listener
      addListener: ( event, callback ) => {
        if ( event === "focus" ) {
          callback();
        }
      }
    } ),
    useRoute: () => ( {
      params: {
        region: {
          latitude: 42.1234567,
          longitude: 42.1234567
        },
        id: "some_id",
        seenDate: "some_date"
      }
    } )
  };
} );

jest.mock( "../../../../../utility/locationHelpers", () => ( {
  __esModule: true,
  fetchTruncatedUserLocation: ( ) => {
    return new Promise( ( resolve ) => {
      resolve( {
        latitude: 42.42,
        longitude: 42.42
      } );
    } );
  }
} ) );

const containerID = "range-map-container";
const mapID = "range-map";
const buttonID = "user-location-button";

describe( "RangeMap", () => {
  test( "should render correctly", () => {
    render( <RangeMap /> );
    // renders the map
    screen.findByTestId( containerID );
    // Create snapshot
    expect( screen ).toMatchSnapshot();
  } );

  test( "should render the map with region as given in props", async () => {
    render( <RangeMap /> );
    // renders the map container
    screen.findByTestId( containerID );

    // renders the map with location passed in navigation props
    await screen.findByTestId( mapID );
    const map = screen.getByTestId( mapID );

    // Expect location given by props
    expect( map.props.region.latitude ).toBe( 42.1234567 );
    expect( map.props.region.longitude ).toBe( 42.1234567 );
  } );

  test( "should change map region to user's location when button is pressed", async () => {
    render( <RangeMap /> );
    screen.findByTestId( containerID );

    // renders the user location button
    await screen.findByTestId( buttonID );
    const locationButton = screen.getByTestId( buttonID );
    expect( locationButton ).toBeTruthy();

    // Create snapshot
    expect( screen ).toMatchSnapshot();

    // Press the location button and expect the map to update
    // TODO: thios button press takes forever, the fireEvent is triggering the onPress but the map is not updating, but taking forever to finish the test
    // fireEvent.press( locationButton );
    // const map = screen.getByTestId( mapID );
    // expect( map.props.region.latitude ).toBe( 42.42 );
  } );
} );
