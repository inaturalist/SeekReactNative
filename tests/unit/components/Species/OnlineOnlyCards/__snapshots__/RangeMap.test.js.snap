// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeMap should change map region to user's location when button is pressed 1`] = `
<RNCSafeAreaView
  edges={
    [
      "top",
    ]
  }
  style={
    {
      "backgroundColor": "#38976d",
      "flex": 1,
    }
  }
  testID="range-map-container"
>
  <View
    style={
      [
        {
          "backgroundColor": "#38976d",
          "flexDirection": "row",
          "paddingBottom": 18,
          "paddingTop": 20.5,
        },
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      accessibilityLabel="Go back to previous screen"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "left": 0,
          "opacity": 1,
          "paddingHorizontal": 23,
          "paddingVertical": 18,
          "position": "absolute",
          "transform": [
            {
              "scaleX": 1,
            },
          ],
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/icons/back.webp",
          }
        }
        style={false}
        tintColor={false}
      />
    </View>
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "maxWidth": 650,
            "paddingTop": 5,
          },
        ]
      }
    >
      RANGE MAP
    </Text>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    useNativeDriverForBackdrop={true}
    visible={false}
  />
  <AIRMap
    initialRegion={null}
    onChange={[Function]}
    onMapReady={[Function]}
    onRegionChangeComplete={[Function]}
    region={
      {
        "latitude": 42.1234567,
        "longitude": 42.1234567,
      }
    }
    style={
      {
        "height": 1259,
        "width": 750,
      }
    }
    testID="range-map"
    zoomEnabled={true}
  >
    <AIRMapUrlTile
      tileSize={512}
      urlTemplate="https://api.inaturalist.org/v1/grid/{z}/{x}/{y}.png?taxon_id=some_id&color=%2377B300&verifiable=true"
    />
    <AIRMapMarker
      coordinate={
        {
          "latitude": 42.1234567,
          "longitude": 42.1234567,
        }
      }
      onPress={[Function]}
      style={
        [
          {
            "backgroundColor": "transparent",
            "position": "absolute",
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/icons/icon-camerabutton-onmap.webp",
          }
        }
      />
    </AIRMapMarker>
    <AIRMapMarker
      coordinate={
        {
          "latitude": 42.42,
          "longitude": 42.42,
        }
      }
      onPress={[Function]}
      style={
        [
          {
            "backgroundColor": "transparent",
            "position": "absolute",
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/icons/icon-location-onmap.webp",
          }
        }
      />
    </AIRMapMarker>
  </AIRMap>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#38976d",
        "borderTopRightRadius": 40,
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "paddingBottom": 23,
        "paddingLeft": 22,
        "paddingRight": 28,
        "position": "absolute",
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "fontFamily": "Whitney-Semibold",
            "fontSize": 19,
            "letterSpacing": 1.12,
            "marginTop": 18,
            "textAlign": "center",
          },
        ]
      }
    >
      LEGEND
    </Text>
  </View>
  <View
    accessibilityLabel="Return to your current location"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "bottom": 19,
        "opacity": 1,
        "position": "absolute",
        "right": 19,
      }
    }
    testID="user-location-button"
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/icon-locationindicator.webp",
        }
      }
    />
  </View>
</RNCSafeAreaView>
`;

exports[`RangeMap should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    [
      "top",
    ]
  }
  style={
    {
      "backgroundColor": "#38976d",
      "flex": 1,
    }
  }
  testID="range-map-container"
>
  <View
    style={
      [
        {
          "backgroundColor": "#38976d",
          "flexDirection": "row",
          "paddingBottom": 18,
          "paddingTop": 20.5,
        },
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      accessibilityLabel="Go back to previous screen"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "left": 0,
          "opacity": 1,
          "paddingHorizontal": 23,
          "paddingVertical": 18,
          "position": "absolute",
          "transform": [
            {
              "scaleX": 1,
            },
          ],
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/icons/back.webp",
          }
        }
        style={false}
        tintColor={false}
      />
    </View>
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "maxWidth": 650,
            "paddingTop": 5,
          },
        ]
      }
    >
      RANGE MAP
    </Text>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    useNativeDriverForBackdrop={true}
    visible={false}
  />
  <AIRMap
    initialRegion={null}
    onChange={[Function]}
    onMapReady={[Function]}
    onRegionChangeComplete={[Function]}
    region={
      {
        "latitude": 42.1234567,
        "longitude": 42.1234567,
      }
    }
    style={
      {
        "height": 1259,
        "width": 750,
      }
    }
    testID="range-map"
    zoomEnabled={true}
  >
    <AIRMapUrlTile
      tileSize={512}
      urlTemplate="https://api.inaturalist.org/v1/grid/{z}/{x}/{y}.png?taxon_id=some_id&color=%2377B300&verifiable=true"
    />
    <AIRMapMarker
      coordinate={
        {
          "latitude": 42.1234567,
          "longitude": 42.1234567,
        }
      }
      onPress={[Function]}
      style={
        [
          {
            "backgroundColor": "transparent",
            "position": "absolute",
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/icons/icon-camerabutton-onmap.webp",
          }
        }
      />
    </AIRMapMarker>
  </AIRMap>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#38976d",
        "borderTopRightRadius": 40,
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "paddingBottom": 23,
        "paddingLeft": 22,
        "paddingRight": 28,
        "position": "absolute",
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "fontFamily": "Whitney-Semibold",
            "fontSize": 19,
            "letterSpacing": 1.12,
            "marginTop": 18,
            "textAlign": "center",
          },
        ]
      }
    >
      LEGEND
    </Text>
  </View>
</RNCSafeAreaView>
`;
